include/libhal-soft/bit_bang_i2c.hpp

```diff
@@ -0,0 +1,177 @@
+   * stretching to ensure the peripheral doesn't hold the line too long
+   *
+   * @throws hal::io_error when the data written to the bus was not
+   * acknowledged
+   */
+  void write(std::span<const hal::byte> p_data_out,
+             function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief This function will handle writing a singular byte each call while
+   * also retrieving the acknowledge bits
+   *
+   * @param p_byte_to_write This is the byte that will be written to the bus
+   * @param p_timeout A timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   *
+   * @return bool true when the byte written was ack'd and false when it was
+   * nack'd
+   */
+  bool write_byte(hal::byte p_byte_to_write,
+                  function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief This function will write a single bit at a time, dealing with
+   * simulating the clock and the clock stretching feature
+   *
+   * @param p_bit_to_write The bit which will be written on the bus
+   * @param p_timeout A timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   */
+  void write_bit(hal::byte p_bit_to_write,
+                 function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief This function will read in as many bytes as allocated inside of the
+   * span while also acking or nacking the data
+   *
+   * @param p_data_in A span which will be filled with the bytes that will be
+   * read from the bus
+   * @param p_timeout A timeout function which is primarily used for clock
+   * stretching to ensure the peripheral doesn't hold the line too long
+   */
+  void read(std::span<hal::byte> p_data_in,
+            function_ref<hal::timeout_function> p_timeout);
+
+  /**
+   * @brief This function is responsible for reading a byte at a time off the
+   * bus and also creating the byte from bits
+   *
+   * @return hal::byte the byte that has been read off of the bus
+   */
+  hal::byte read_byte();
+
+  /**
+   * @brief This function is responsible for reading a single bit at a time. It
+   * will manage the clock and will release sda (pull it high) every time it is
+   * called
+   *
+   * @return hal::byte which will house the single byte read from the bus
+   */
+  hal::byte read_bit();
+
+  /// @brief An output pin which is the i2c scl pin
+  output_pin* m_scl;
+
+  /// @brief An output pin which is the i2c sda pin
+  output_pin* m_sda;
+
+  /// @brief A steady_clock provides a mechanism to delay the clock pulses of
+  /// the scl line.
+  steady_clock* m_clock;
+
+  /// @brief The time that scl will be held high for
+  uint64_t m_scl_high_ticks;
+
+  /// @brief The time that scl will be held low for
+  uint64_t m_scl_low_ticks;
+
+  /// @brief All the information that the bus will need to operate on
```

Comment:
```suggestion
  /// @brief All the information that the bus will need to operate on
```
Update comment and explain that its used by `driver_configure()` so it can preserve the duty cycle passed to the constructor.